src/v/redpanda/application.cc

```diff
@@ -1716,3 +1723,70 @@ void application::start_kafka(
       "Started Kafka API server listening at {}",
       config::node().kafka_api());
 }
+
+/**
+ * Feature table is generally updated via controller, but we need it to
+ * be initialized very early in startup so that other subsystems (including
+ * e.g. the controller raft group) may rely on up to date knowledge of which
+ * feature bits are enabled.
+ */
+void application::load_feature_table_snapshot() {
+    auto val_bytes_opt = storage.local().kvs().get(
+      storage::kvstore::key_space::controller,
+      features::feature_table_snapshot::kvstore_key());
+
+    if (!val_bytes_opt) {
+        // No snapshot?  Probably we are yet to join cluster.
+        return;
+    }
+
+    features::feature_table_snapshot snap;
+    try {
+        snap = serde::from_iobuf<features::feature_table_snapshot>(
+          std::move(*val_bytes_opt));
+    } catch (...) {
+        // Do not block redpanda from starting if there is something invalid
+        // here: the feature table should get replayed eventually via
+        // the controller.
+        vlog(
+          _log.error,
+          "Exception decoding feature table snapshot: {}",
+          std::current_exception());
+#ifndef NDEBUG
+        vassert(false, "Snapshot decode failed");
+#endif
+    }
+
+    auto my_version = features::feature_table::get_latest_logical_version();
+    if (my_version < snap.version) {
```

Comment:
what's the behavior here when `snap` is default initialized and we fall through from the catch above?

Reply:
That is a mistake, the catch above should have returned.  Fixed.